### ISSUE TITLE
Bogart failure (during assemble task, contig construction)

### ISSUE THREAD

**Issue by lubos22** [marbl/canu#1958]
Dear Canu team,
I am very grateful for the software, thank you! Just now I am however facing a crash at the very end of the pipeline (and after several weeks running). I am assembling Nanopore raw reads of ca. 80 Gb (genome size 0.9 Gb) in the three top-level tasks. After successful correcting and trimming the program failed in the end of assembling (Bogart). I have tried to resume several times, also in other versions of Canu including a grid installation where Canu always runs well. No success.
I would be very grateful for any help,
Lubomir
 
`Command:`
canu \
  useGrid=false \
  -p len3000 \
  -d len3000-erate-0.13 \
  genomeSize=0.9g \
  correctedErrorRate=0.13 \
  -trimmed -corrected -nanopore len3000/len3000.trimmedReads.fasta.gz

`Version:`
canu snapshot v2.2-development +79 changes

`System:`
Ubuntu 18
stand-alone PC; AMD Threadripper 2970WX 24(48)x3.0; RAM 64 GB, HardDisc 16 TB

`unitigger.err file content:`
```
==> PARAMETERS.

Resources:
  Memory                63 GB
  Compute Threads       16 (command line)

Lengths:
  Minimum read          0 bases
  Maximum read          4294967295 bases
  Minimum overlap       500 bases

Overlap Error Rates:
  Graph                 0.130 (13.000%)
  Max                   0.130 (13.000%)

Deviations:
  Graph                 12.000
  Bubble                1.000
  Repeat                1.000

Similarity Thresholds:
  Graph                 0.000
  Bubble                0.010
  Repeat                0.010

Edge Confusion:
  Absolute              2500
  Percent               15.0000

Unitig Construction:
  Minimum intersection  500 bases
  Maxiumum placements   2 positions

Debugging Enabled:
  (none)

==> LOADING AND FILTERING OVERLAPS.

ReadInfo()-- Found     1168816 reads.

OverlapCache()-- limited to 64512MB memory (user supplied).

OverlapCache()--       8MB for read data.
OverlapCache()--      35MB for best edges.
OverlapCache()--     115MB for tigs.
OverlapCache()--      31MB for tigs - read layouts.
OverlapCache()--      44MB for tigs - error profiles.
OverlapCache()--   16128MB for tigs - error profile overlaps.
OverlapCache()--    6434MB for other processes.
OverlapCache()-- ---------
OverlapCache()--   22821MB for data structures (sum of above).
OverlapCache()-- ---------
OverlapCache()--      22MB for overlap store structure.
OverlapCache()--   41668MB for overlap data.
OverlapCache()-- ---------
OverlapCache()--   64512MB allowed.
OverlapCache()--
OverlapCache()-- Retain at least 77 overlaps/read, based on 38.92x coverage.
OverlapCache()-- Initial guess at 2336 overlaps/read.
OverlapCache()--
OverlapCache()-- Adjusting for sparse overlaps.
OverlapCache()--
OverlapCache()--               reads loading olaps          olaps               memory
OverlapCache()--   olaps/read       all      some          loaded                 free
OverlapCache()--   ----------   -------   -------     ----------- -------     --------
OverlapCache()--         2336   1145467     23349       194721309  75.06%      38697 MB
OverlapCache()--       110951   1168816         0       259408126 100.00%      37710 MB
OverlapCache()--
OverlapCache()-- Loading overlaps.
OverlapCache()--
OverlapCache()--          read from store           saved in cache
OverlapCache()--   ------------ ---------   ------------ ---------
OverlapCache()--       26336065 (010.15%)       26206891 (010.10%)
OverlapCache()--       49023054 (018.90%)       48824718 (018.82%)
OverlapCache()--       73211543 (028.22%)       72819760 (028.07%)
OverlapCache()--       99710294 (038.44%)       98913930 (038.13%)
OverlapCache()--      120369813 (046.40%)      119530438 (046.08%)
OverlapCache()--      141929632 (054.71%)      141048017 (054.37%)
OverlapCache()--      168020507 (064.77%)      167085805 (064.41%)
OverlapCache()--      188349292 (072.61%)      187391416 (072.24%)
OverlapCache()--      209112621 (080.61%)      208124985 (080.23%)
OverlapCache()--      228492612 (088.08%)      227484050 (087.69%)
OverlapCache()--      247342050 (095.35%)      246316092 (094.95%)
OverlapCache()--   ------------ ---------   ------------ ---------
OverlapCache()--      259408126 (100.00%)      258364078 (099.60%)
OverlapCache()--
OverlapCache()-- Ignored 908986 duplicate overlaps.
OverlapCache()--
OverlapCache()-- Symmetrizing overlaps.
OverlapCache()--   Finding missing twins.
OverlapCache()--   In 258364078 overlaps:
OverlapCache()--     Found 36265590 overlaps with non-symmetric error rates.
OverlapCache()--     Found 29836 overlaps with missing twins.
OverlapCache()--     Dropped 0 weak missing-twin overlaps.
OverlapCache()--   Shifting overlaps.
OverlapCache()--   Adding missing twins.
OverlapCache()--   Sorting overlaps.
OverlapCache()--   Checking overlap symmetry.
OverlapCache()--   Finished.

BestOverlapGraph()-- Computing Best Overlap Graph.
BestOverlapGraph()-- Allocating best edges (35MB).
BestOverlapGraph()-- Filtering high error edges.
BestOverlapGraph()--   Ignore overlaps with more than 1.747626% error.
BestOverlapGraph()-- Filtering reads with a gap in overlap coverage.
BestOverlapGraph()--   23352 reads removed.
BestOverlapGraph()-- Filtering reads with lopsided best edges (more than 25.00% different).
BestOverlapGraph()--   985 reads have lopsided edges.
BestOverlapGraph()-- Filtering spur reads.
BestOverlapGraph()--   Initial          24970 terminal  spur reads -    25460/25465    5'/3' spur path reads.
BestOverlapGraph()--   Iteration 1 -    24970 terminal  spur reads -    25336/25351    5'/3' spur path reads -      444/398      edges changed to avoid spur path.
BestOverlapGraph()--   Iteration 2 -    24970 terminal  spur reads -    25290/25318    5'/3' spur path reads -       58/37       edges changed to avoid spur path.
BestOverlapGraph()--   Iteration 3 -    24970 terminal  spur reads -    25289/25314    5'/3' spur path reads -        3/2        edges changed to avoid spur path.
BestOverlapGraph()--   Iteration 4 -    24970 terminal  spur reads -    25289/25314    5'/3' spur path reads -        0/0        edges changed to avoid spur path.
BestOverlapGraph()--   Final            26902 confirmed spur reads -    25289/25314    5'/3' spur path reads.

==> BUILDING GREEDY TIGS.

breakSingletonTigs()-- Removed 7 singleton tigs; reads are now unplaced.
optimizePositions()-- Optimizing read positions for 1168817 reads in 2701 tigs, with 16 threads.
optimizePositions()--   Allocating scratch space for 1168817 reads (36525 KB).
optimizePositions()--   Initializing positions with 16 threads.
optimizePositions()--   Recomputing positions, iteration 1, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1164046 reads
optimizePositions()--     changed:     4771 reads
optimizePositions()--   Recomputing positions, iteration 2, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1166935 reads
optimizePositions()--     changed:     1882 reads
optimizePositions()--   Recomputing positions, iteration 3, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1168715 reads
optimizePositions()--     changed:      102 reads
optimizePositions()--   Recomputing positions, iteration 4, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1168765 reads
optimizePositions()--     changed:       52 reads
optimizePositions()--   Recomputing positions, iteration 5, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1168788 reads
optimizePositions()--     changed:       29 reads
optimizePositions()--   Expanding short reads with 16 threads.
optimizePositions()--   Updating positions.
optimizePositions()--   Finished.
splitDiscontinuous()-- Tested 2693 tigs, split none.
detectSpur() done.
tested          327
nEdges      5'  231   3'  239
nPotential       16        13
nVerified         3         3

==> PLACE CONTAINED READS.

computeErrorProfiles()-- Computing error profiles for 2701 tigs, with 16 threads.
computeErrorProfiles()-- Finished.

placeContains()-- placing 1049310 contained and 50499 unplaced reads, with 16 threads.
placeContains()-- Placed 1037565 contained reads and 21799 unplaced reads.
placeContains()-- Failed to place 11745 contained reads (too high error suspected) and 28700 unplaced reads (lack of overlaps suspected).
optimizePositions()-- Optimizing read positions for 1168817 reads in 2701 tigs, with 16 threads.
optimizePositions()--   Allocating scratch space for 1168817 reads (36525 KB).
optimizePositions()--   Initializing positions with 16 threads.
optimizePositions()--   Recomputing positions, iteration 1, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 741757 reads
optimizePositions()--     changed:   427060 reads
optimizePositions()--   Recomputing positions, iteration 2, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1096212 reads
optimizePositions()--     changed:    72605 reads
optimizePositions()--   Recomputing positions, iteration 3, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1125526 reads
optimizePositions()--     changed:    43291 reads
optimizePositions()--   Recomputing positions, iteration 4, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1126353 reads
optimizePositions()--     changed:    42464 reads
optimizePositions()--   Recomputing positions, iteration 5, with 16 threads.
optimizePositions()--     Reset zero.
optimizePositions()--     Checking convergence.
optimizePositions()--     converged: 1128838 reads
optimizePositions()--     changed:    39979 reads
optimizePositions()--   Expanding short reads with 16 threads.
optimizePositions()--   Updating positions.
optimizePositions()--   Finished.
splitDiscontinuous()-- Tested 2693 tigs, split none.

==> MERGE ORPHANS.

computeErrorProfiles()-- Computing error profiles for 2701 tigs, with 16 threads.
computeErrorProfiles()-- Finished.

findPotentialOrphans()-- working on 2701 tigs.
findPotentialOrphans()-- found 2548 potential orphans.
mergeOrphans()-- flagged    1145        bubble tigs with 72359 reads
mergeOrphans()-- placed      322 unique orphan tigs with 10458 reads
mergeOrphans()-- shattered     8 repeat orphan tigs with 98 reads
mergeOrphans()-- ignored       8               tigs with 5053 reads; failed to place
mergeOrphans()--

==> MARK SIMPLE BUBBLES.
    using 0.010000 user-specified threshold


findPotentialOrphans()-- working on 2701 tigs.
findPotentialOrphans()-- found 2317 potential orphans.
bogart: bogart/AS_BAT_PlaceReadUsingOverlaps.C:189: void placeRead_fromOverlaps(TigVector&, Unitig*, uint32, uint32, double, uint32, BAToverlap*, uint32&, overlapPlacement*): Assertion `op.position.isForward() == op.verified.isForward()' failed.

Failed with 'Aborted'; backtrace (libbacktrace):
utility/src/utility/system-stackTrace.C::83 in _Z17AS_UTL_catchCrashiP7siginfoPv()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
bogart/AS_BAT_PlaceReadUsingOverlaps.C::189 in _Z22placeRead_fromOverlapsR9TigVectorP6UnitigjjdjP10BAToverlapRjP16overlapPlacement()
bogart/AS_BAT_PlaceReadUsingOverlaps.C::564 in _Z22placeReadUsingOverlapsR9TigVectorP6UnitigjRSt6vectorI16overlapPlacementSaIS4_EEjd()
bogart/AS_BAT_MergeOrphans.C::400 in _Z24findOrphanReadPlacementsR9TigVectorRSt3mapIjSt6vectorIjSaIjEESt4lessIjESaISt4pairIKjS4_EEEdddb._omp_fn.0()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
Aborted (core dumped)
```

**Comment by brianwalenz** [comment]
I can't recall having seen that particular error before.

Since you're already using a non-release version, you might as well go ahead and update canu to the master tip (e.g., `git pull`).  I don't see anything directly related to this in the changes since (approximately) the version you have though.

If that fails, the easiest way forward will be for you to send us the data to reproduce the crash: `unitigging/len3000.ovlStore` and `len3000.seqStore` (but NOT including the large `blobs` files).  Note that the sequence data is stored in the blobs files.

**Comment by skoren** [comment]
Idle, hopefully addressed by the 2.2 release, feel free to re-open if not.
